### PR TITLE
Don't fear the (sub)reaper

### PR DESCRIPTION
**Pull request by mudler**
This is still a wip, and commit message will be replaced :) - needs also new RWP release as sessions were introduced recently - and there was a typo preventing orphaned child resolution during event firing when using subreapers, not in the feature itself (and update cpanfile accordingly once its done).

### CODE REVIEW

**Comment by foursixnine** [comment]
https://www.youtube.com/watch?v=ClQcUyhoxTg

**Comment by mudler** [comment]
More cowbells! https://youtu.be/Gwb4o2qs7Pc?t=14

File: lib/OpenQA/Worker/Jobs.pm [hunk 1]
```diff
@@ -544,20 +543,18 @@ sub start_job {
...
-                if ($status != 0) {
-                    _stop_job_2('dead_children', $job->{id});
-                }
-            });
-    }
 
+    $worker->{child}->on(
+        collected => sub {
+            my $self = shift;
+            STDERR->printflush("collect status: " . $self->exit_status . "\n");
```
Comment (foursixnine) [inline lib/OpenQA/Worker/Jobs.pm]: Will this end in the logs?

File: lib/OpenQA/Worker/Jobs.pm [hunk 2]
```diff
@@ -544,20 +543,18 @@ sub start_job {
...
+            # TODO: deal when the job was restarted. The process is always killed.
+            # This is only problematic if we die, Otherwise this is just to exterminate
```
Comment (foursixnine) [inline lib/OpenQA/Worker/Jobs.pm]: Now that you are at this, mind changing the wording to process tree?

File: lib/OpenQA/Worker/Jobs.pm (hunk 1 above)
Comment (mudler) [reply lib/OpenQA/Worker/Jobs.pm]: this changes are not going so much distant from https://github.com/os-autoinst/openQA/pull/1579

File: lib/OpenQA/Worker/Jobs.pm [hunk 3]
```diff
@@ -914,8 +902,10 @@ sub check_backend {
     log_debug("checking backend state");
 
     return log_debug("backend is running") if $worker->{child}->is_running();
+    return log_debug("backend state not known") unless defined $worker->{child}->exit_status;
```
Comment (foursixnine) [inline lib/OpenQA/Worker/Jobs.pm]: I would move this to log_error

File: lib/OpenQA/Worker/Jobs.pm [hunk 4]
```diff
@@ -914,8 +902,10 @@ sub check_backend {
...
 
-    if ($worker->{child}->is_running()) {
+    log_debug("backend is not running anymore");
```
Comment (foursixnine) [inline lib/OpenQA/Worker/Jobs.pm]: This one to info

**Comment by coolo** [comment]
with RWP 0.16 the worker is *always* detecting dead children - as collect_status finds the children. There seems to be an incompatible change in there :(

Had to notice after reinstalling the arm worker

**Comment by mudler** [comment]
So yes, let me be clear a bit https://github.com/os-autoinst/openQA/pull/1579 was a premature merge - as collect_status was an internal event and you was required to manually check what you are collecting for - aside from the missing bit shifting to be sure of what is the real status of the process. 
The workers as it is in the current code are collecting all statuses from all processes.

in >0.16 there is the 'collected' event that it's fired when we have the exit status of the process.

File: lib/OpenQA/Worker/Jobs.pm (hunk 3 above)
Comment (mudler) [reply lib/OpenQA/Worker/Jobs.pm]: actually we could remove that timer now, if stop_job would not be so 'blocking' - give me some time to try to re-elaborate this

**Comment by mudler** [comment]
https://build.opensuse.org/request/show/580145

**Comment by mudler** [comment]
tweaked MAX_JOB_TIME: http://e122.suse.de/tests/7790/file/autoinst-log.txt

standard jobs:
http://e122.suse.de/tests/7796
http://e122.suse.de/tests/7794
http://e122.suse.de/tests/7792
http://e122.suse.de/tests/7795
http://e122.suse.de/tests/7799
http://e122.suse.de/tests/7797
http://e122.suse.de/tests/7798
http://e122.suse.de/tests/7800
http://e122.suse.de/tests/7801
http://e122.suse.de/tests/7802
http://e122.suse.de/tests/7803
http://e122.suse.de/tests/7804
http://e122.suse.de/tests/7805
http://e122.suse.de/tests/7806
http://e122.suse.de/tests/7807
http://e122.suse.de/tests/7808

**Comment by foursixnine** [comment]
giphy-downsized

**Pull request merged**